match: nonlinear pat refs in list-no-order

AlexKnauth opened the pull request:
Allows patterns inside `list-no-order` to refer to nonlinear patterns declared prior, such as in `(cons a (list-no-order a rst ...))`.

Includes nonlinear patterns declared under ellipses, such as in `(list (cons a (list-no-order a rst ...)) ...)`.

This also makes nonlinear pattern declarations syntax errors if the first use is in the scope of a `...` pattern that doesn't encompass all uses.

AlexKnauth committed:
match: nonlinear pat refs in list-no-order

only works when not inside ellipses so far

AlexKnauth committed:
match: ellipses nonlinear refs in list-no-order

samth commented on racket/collects/racket/match/compiler.rkt:
```diff
@@ -551,5 +543,27 @@
+       (if (identifier? id)
+           id
+           (begin
+             (log-error "non-linear pattern used in `match` with ... at ~a and ~a"
```
If this is supposed to work now this log should be removed.

AlexKnauth replied:
I didn't want things like this to change for now:
```racket
(match '(1 2 3 4)
  [(list a ... a) a])
```
Order matters. When the `a` without ellipses comes first, that's when it works as a nonlinear pattern:
```racket
(match '(1 1 1 1)
  [(list a a ...) a])
```

AlexKnauth replied:
I've updated the PR, it now makes nonlinear pattern declarations syntax errors if the first use is in the scope of a `...` pattern that doesn't encompass all uses.

AlexKnauth commented:
I have an idea for checking the problems that lead to warnings:

Track ellipsis depth, and when merging pattern-variables together into non-linear patterns, only accept them when the ellipsis-depth of the first one, the declaration, is less than or equal to the ellipsis-depth of the later ones, the references.

So `(list a a ...)` would be allowed because the declaration ellipsis-depth 0 is less than the reference ellipsis-depth 1, but `(list a ... a)` would be an error because the declaration ellipsis-depth 1 is greater than the reference ellipsis-depth 0.

AlexKnauth committed:
Add 'match-ellipsis-depth syntax property

AlexKnauth committed:
match: order seen left-to-right

nonlinear pattern declarations on the left, references on the right

AlexKnauth committed:
match: fix depth of (list id ...) as id

AlexKnauth committed:
match: test nonlinear pattern reference in id ...

AlexKnauth committed:
match: Error on non-linear decl with ... wrt ref

doesn't really catch it yet because of the renaming affects "seen" and not just final binding

AlexKnauth committed:
match: Refactor bound-vars, orig/rename versions

AlexKnauth committed:
match: nonlinear orig/non-renamed vars for seen

for detecting nonlinear patterns, use the original vars in the seen list, not their renamed versions

AlexKnauth committed:
match: test error on nonlinear pat decl with ...

AlexKnauth committed:
match: nonlinear pat decls zero depth wrt refs

fix find-seen to check that nonlinear pattern declarations have zero ellipsis-depth with respect to their references

AlexKnauth committed:
match: document error on nonlinear pat decl in ...

AlexKnauth committed:
match: test nonlinear ... error with non-null tail

samth commented:
Closing since #5467 changes things in a different direction.
